File: frontend/src/api/bikes.ts
import { apiRequest } from "./client";

// ─── Tipos ──────────────────────────────────────────────────────────────────
export interface BikeOut {
  id: number;
  code: string;
  status: "available" | "in_use" | "maintenance";
  station_id: number | null;
  latitude: number | null;
  longitude: number | null;
  total_km: number;
  max_km: number;
  last_maintenance_km: number;
  needs_maintenance: boolean;
  qr_path: string | null;
}

export interface BikeCreate {
  status?: string;
  station_id?: number;
  latitude?: number;
  longitude?: number;
  max_km?: number;
}

export interface BikeUpdate {
  status?: string;
  station_id?: number;
  latitude?: number;
  longitude?: number;
  total_km?: number;
  max_km?: number;
  last_maintenance_km?: number;
  needs_maintenance?: boolean;
}

export interface BikeHistoryItem {
  ride_id: number;
  user: { id: number; full_name: string; email: string };
  start_time: string | null;
  end_time: string | null;
  distance_km: number | null;
  start_station_id: number | null;
  end_station_id: number | null;
}

export interface BikeHistoryOut {
  bike_id: number;
  bike_code: string;
  total_km: number;
  total_rides: number;
  rides: BikeHistoryItem[];
}

// ─── API calls (todos requieren autenticación) ───────────────────────────────
export const getBikes = () =>
  apiRequest<BikeOut[]>("/bikes", { auth: true });

export const getBike = (id: number) =>
  apiRequest<BikeOut>(`/bikes/${id}`, { auth: true });

export const createBike = (data: BikeCreate) =>
  apiRequest<BikeOut>("/bikes", { method: "POST", body: JSON.stringify(data), auth: true });

export const updateBike = (id: number, data: BikeUpdate) =>
  apiRequest<BikeOut>(`/bikes/${id}`, { method: "PUT", body: JSON.stringify(data), auth: true });

export const deleteBike = (id: number) =>
  apiRequest<{ message: string }>(`/bikes/${id}`, { method: "DELETE", auth: true });

export const getBikeQrUrl = (id: number) => `/bikes/${id}/qr`;

export const getBikeHistory = (id: number) =>
  apiRequest<BikeHistoryOut>(`/bikes/${id}/history`, { auth: true });

File: backend/schemas/bikes.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class BikeCreate(BaseModel):
    """Sólo se pasa el status inicial; el código UUID y el QR se autogeneran."""
    status: Optional[str] = "available"
    station_id: Optional[int] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    max_km: Optional[float] = 500


class BikeUpdate(BaseModel):
    status: Optional[str] = None
    station_id: Optional[int] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    total_km: Optional[float] = None
    max_km: Optional[float] = None
    last_maintenance_km: Optional[float] = None
    needs_maintenance: Optional[bool] = None


class BikeOut(BaseModel):
    id: int
    code: str
    status: str
    station_id: Optional[int]
    latitude: Optional[float]
    longitude: Optional[float]
    total_km: float
    max_km: float
    last_maintenance_km: float
    needs_maintenance: bool
    qr_path: Optional[str]

    class Config:
        from_attributes = True


class MaintenanceCreate(BaseModel):
    bike_id: int
    description: Optional[str] = None
    km_at_service: Optional[float] = None


class MaintenanceOut(BaseModel):
    id: int
    bike_id: int
    date: Optional[datetime]
    description: Optional[str]
    km_at_service: Optional[float]

    class Config:
        from_attributes = True


# ────────────────────────────────────────────────
#  HISTORIAL DE VIAJES
# ────────────────────────────────────────────────
class RideUserInfo(BaseModel):
    id: int
    full_name: str
    email: str

    class Config:
        from_attributes = True


class BikeRideHistoryItem(BaseModel):
    ride_id: int
    user: RideUserInfo
    start_time: Optional[datetime]
    end_time: Optional[datetime]
    distance_km: Optional[float]
    start_station_id: Optional[int]
    end_station_id: Optional[int]

    class Config:
        from_attributes = True


class BikeHistoryOut(BaseModel):
    bike_id: int
    bike_code: str
    total_km: float
    total_rides: int
    rides: List[BikeRideHistoryItem]
